File: stockagent/portfolio_plans.py
"""Category-first cash allocation; output is a proposal, never a ledger trade."""
from datetime import date
from decimal import Decimal, ROUND_DOWN

from .portfolio_ledger import ZERO, number, scalar, snapshot


def proposals(portfolio, today=None):
    today = today or date.today().isoformat()
    summary = snapshot(portfolio)
    results = []
    for plan in portfolio.get('plans', []):
        if plan.get('enabled') is False:
            continue
        account = next(a for a in summary['accounts'] if a['id'] == plan['account_id'])
        budget = number(plan['amount'])
        result = {'id': plan['id'], 'name': plan['name'], 'date': plan['next_date'], 'due': plan['next_date'] <= today,
                  'budget': scalar(budget), 'orders': [], 'blocked': [], 'unallocated': scalar(budget)}
        if account['cash'] is None:
            result['blocked'].append('请先核对账户可用现金')
            results.append(result)
            continue
        budget = min(budget, number(account['cash']))
        if plan['strategy'] == 'initial' and summary['total']['value'] is None:
            result['blocked'].append('持仓行情不完整，无法计算建仓缺口')
            results.append(result)
            continue
        allocations = []
        for category in summary['categories']:
            if plan['strategy'] == 'initial':
                weight = max(ZERO, number(plan['target_capital']) * number(category['target_pct']) / 100 - number(category['value']))
            else:
                weight = number(category['target_pct'])
            weighted = [r for r in portfolio['products'] if r['category_id'] == category['id'] and r.get('active',True) and r.get('allocation_weight',0)>0]
            if weighted:
                weight_total = sum(number(r['allocation_weight']) for r in weighted)
                allocations.extend((category,weight*number(r['allocation_weight'])/weight_total,r) for r in weighted)
            else:
                allocations.append((category, weight, None))
        total_weight = sum((weight for _, weight, _ in allocations), ZERO)
        if not total_weight:
            result['blocked'].append('当前没有配置缺口')
            results.append(result)
            continue
        spent = ZERO
        for category, weight, designated in allocations:
            if not weight:
                continue
            allocation = budget * weight / total_weight
            if plan['strategy'] == 'initial':
                allocation = min(allocation, weight)
            candidates = [p for p in portfolio['products'] if p['category_id'] == category['id'] and p.get('active', True)]
            primary = designated or next((p for p in candidates if p['id'] == category.get('primary_product_id')), None)
            if not primary:
                result['blocked'].append(f"{category['name']}：请指定主要买入产品")
                continue
            ordered = [primary] + ([p for p in candidates if p['id'] != primary['id']] if category.get('allow_substitution') and not designated else [])
            selected = None
            reason = ''
            for product in ordered:
                mark = product.get('mark')
                if not mark or not 0 <= (date.fromisoformat(today) - date.fromisoformat(mark['date'])).days <= 7:
                    reason = '缺少近期价格／净值'
                    continue
                if product.get('purchase_blocked'):
                    reason = '产品暂停买入'
                    continue
                if plan['strategy'] == 'dip':
                    history = sorted((r for r in product.get('price_history', []) if r['date'] <= today),key=lambda r:r['date'])
                    if len(history) < 20:
                        reason = '回撤观察不足20个价格记录'
                        continue
                    peak = max(number(r['price']) for r in history[-120:])
                    drawdown = (1 - number(mark['price']) / peak) * 100
                    if drawdown < number(plan['drawdown_pct']):
                        reason = f"未达到{plan['drawdown_pct']}%回撤"
                        continue
                selected = product
                break
            if not selected:
                result['blocked'].append(f"{category['name']}：{reason}")
                continue
            cost = account.get('trading_cost') or {}
            rate = number(selected.get('fee_rate_pct', cost.get('commission_rate_pct', 0))) / 100
            minimum_fee = number(cost.get('min_commission', 0)) if selected['kind'] == 'exchange' else ZERO
            limit = selected.get('purchase_limit')
            if limit is not None:
                allocation = min(allocation, number(limit))
            price = number(selected['mark']['price'])
            if selected['kind'] == 'exchange':
                lot = number(selected.get('lot_size', 100))
                shares = (max(ZERO, min(allocation / (1 + rate), allocation - minimum_fee)) / price / lot).to_integral_value(rounding=ROUND_DOWN) * lot
                gross = shares * price
                fee = max(minimum_fee, gross * rate) if shares else ZERO
                amount = gross + fee
                if not shares:
                    result['blocked'].append(f"{category['name']}：预算不足一个交易单位及费用")
                    continue
            else:
                amount = allocation.quantize(Decimal('0.01'), rounding=ROUND_DOWN)
                fee = amount - amount / (1 + rate)
                shares = None
                if amount < number(selected.get('min_purchase', 1)):
                    result['blocked'].append(f"{category['name']}：未达到最低申购金额")
                    continue
            spent += amount
            result['orders'].append({'category_id': category['id'], 'product_id': selected['id'], 'account_id': account['id'],
                                     'amount': scalar(amount), 'estimated_fee': scalar(fee), 'shares': scalar(shares),
                                     'price': scalar(price), 'substituted': selected['id'] != primary['id']})
        result['unallocated'] = scalar(number(plan['amount']) - spent)
        results.append(result)
    return results

File: js/views/portfolio-plans.js
import { escapeHtml, escapeAttr } from "../utils.js";
import { simulationPanel } from "./portfolio-simulation.js";

const money = v => v == null ? "未知" : `¥${Number(v).toLocaleString("zh-CN",{maximumFractionDigits:2})}`;
const strategies = {initial:"初期建仓",dca:"周期定投",dip:"逢低加仓"};

function planForm(p, plan = {}) {
  return `<form data-form="plan" data-id="${escapeAttr(plan.id||'')}" class="portfolio-form">
    <label>名称<input name="name" value="${escapeAttr(plan.name||'')}" required></label>
    <label>策略<select name="strategy">${Object.entries(strategies).map(([id,name])=>`<option value="${id}">${name}</option>`).join("")}</select></label>
    <label>资金账户<select name="account_id">${p.accounts.map(a=>`<option value="${escapeAttr(a.id)}">${escapeHtml(a.name)}</option>`).join("")}</select></label>
    <label>本期预算（元）<input name="amount" type="number" min="0.01" step="0.01" value="${Number(plan.amount)||1000}" required></label>
    <label>建仓目标总额（元）<input name="target_capital" type="number" min="0" step="0.01" value="${Number(plan.target_capital)||0}"></label>
    <label>回撤阈值（%）<input name="drawdown_pct" type="number" min="0.1" max="99.9" step="0.1" value="${Number(plan.drawdown_pct)||5}"></label>
    <label>周期<select name="cadence"><option value="monthly">每月</option><option value="weekly">每周</option></select></label>
    <label>下次执行日期<input name="next_date" type="date" value="${escapeAttr(plan.next_date||new Date().toLocaleDateString('sv-SE'))}" required></label>
    <button class="primary-button">保存计划</button></form>`.replace(/<select name="([^"]+)">([\s\S]*?)<\/select>/g,(_,key,content)=>`<select name="${key}">${content.replace(/<option value="([^"]+)"/g,(tag,value)=>tag+(String(plan[key])===value?' selected':''))}</select>`);
}

export function portfolioPlans(p, proposals = []) {
  const legacy = p.migration?.legacy_plan || {};
  return `<section class="portfolio-section"><h2>新增投资计划</h2>${planForm(p,{amount:legacy.amount,target_capital:legacy.capital_base})}</section>
    ${p.plans.map(plan=>{
      const proposal=proposals.find(r=>r.id===plan.id);
      return `<section class="portfolio-section"><div class="portfolio-heading"><h2>${escapeHtml(plan.name)} · ${strategies[plan.strategy]}</h2><button class="ghost-button" ${plan.enabled===false?'disabled':''} data-plan-advance="${escapeAttr(plan.id)}">本期处理完毕</button></div>
        <p>${plan.enabled===false?'已停用 · ':''}${escapeHtml(plan.next_date)} · ${plan.cadence==='weekly'?'每周':'每月'} · 预算 ${money(plan.amount)}${proposal?.due?' · 待执行':''}</p>
        ${proposal ? `<div class="portfolio-table-wrap"><table class="portfolio-table"><thead><tr><th>类别</th><th>产品</th><th>预算金额</th><th>预计费用</th><th>建议份额</th></tr></thead><tbody>${proposal.orders.map(order=>`<tr><th>${escapeHtml(p.categories.find(c=>c.id===order.category_id)?.name||'')}</th><td><button class="link-button" data-product="${escapeAttr(order.product_id)}">${escapeHtml(p.products.find(r=>r.id===order.product_id)?.name||'')}</button>${order.substituted?'（类内替代）':''}</td><td>${money(order.amount)}</td><td>${money(order.estimated_fee)}</td><td>${order.shares??'以申购确认份额为准'}</td></tr>`).join('')}</tbody></table></div><p>保留现金 ${money(proposal.unallocated)}</p>${proposal.blocked.map(reason=>`<p class="portfolio-warning">${escapeHtml(reason)}</p>`).join('')}`:`<p>${plan.enabled===false?'计划已停用':'保存并启用组合后生成分配明细。'}</p>`}
        <button class="primary-button" data-go="transactions">录入实际交易</button>
        <button class="ghost-button" data-plan-toggle="${escapeAttr(plan.id)}">${plan.enabled===false?'启用':'停用'}</button><details><summary>编辑计划</summary>${planForm(p,plan)}</details></section>`;
    }).join('')}
    <details class="portfolio-section"><summary>计算规则</summary><p>初期建仓按类别目标金额减现有市值分配；定投按类别目标比例分配；逢低加仓按主要产品最近120条历史价格的高点回撤判断，至少需要20条记录。未满足条件的资金保留现金。</p><p>计划不会自动产生真实成交。场内按交易单位及费用取整；场外以申购金额生成建议，实际份额在交易记录中确认。</p></details>${simulationPanel(p)}`;
}
